### ISSUE TITLE
Broker accepts connections with clientId longer than 23 characters for MQTT v3.1

### ISSUE THREAD

**Issue by mjacoby** [hivemq/hivemq-community-edition#160]
## Expected behavior
Broker should reject connection with protocol version 3.1 and clientId longer than 23 characters with return code 2 (Identifier Rejected) as defined in the standard https://public.dhe.ibm.com/software/dw/webservices/ws-mqtt/mqtt-v3r1.html#connect
## Actual behavior
Broker accepts connection with return code 0 (Connection Accepted).
### Steps
Open client connection with protocol version 3.1 and a clientId longer than 23 characters, e.g. aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
### Reproducer code

- requires a HiveMQ CE broker running at localhost:1883
- Eclipse Paho Library 1.2.2

```
public static void main(String[] args) {
	MqttConnectOptions options = new MqttConnectOptions();
	options.setMqttVersion(MqttConnectOptions.MQTT_VERSION_3_1);
	try {
		MqttClient client = new MqttClient("tcp://localhost:1883", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", new MemoryPersistence());
		client.connect(options);
		client.disconnect();
		client.close();
	} catch (MqttException ex) {
		System.err.println(ex);
	}
}
```
## Details
- Affected HiveMQ CE version(s): tested with docker container hivemq/hivemq-ce:latest
- Used JVM version:

**Comment by SgtSilvio** [comment]
Hi @mjacoby 
You are right that HiveMQ does not differentiate between MQTT versions regarding the length of the client id. You can set a limit for the client id length in the configuration file via `<restrictions><max-client-id-length>`
In our opinion there is no good reason to restrict this setting particularly for MQTT 3.1.
If you have a reason, please feel free to comment.

**Comment by mjacoby** [comment]
Thanks for the feedback! I did not know about this property in the configuration file.

I understand that you don't see a reason to change this behavior because your server offers even longer clientIds than defined in MQTT v3.1 so one assumes that this should not be an issue.

However, I some scenarios this can lead to problems. For example, I came across a piece of software that connectes to an MQTT server using MQTT v3.1 and a clientId longer than 23 characters. Per standard, this should never work. Unfortunately, it seems that the developers did test against a server that does allow such long clientIds, maybe HiveMQ CE, I can't tell for sure. So when I tried to run that software against another server which enforces the clientId to be at most 23 characters like the standard defines I was not able to run the software.

Of course there are multiple issues here, one of them being the client library also not enforcing that length limit on clientId or the developer using a longer clientId. However, if every library (including HiveMQ CE) would have fully implemented the standard this would not be able to happen.

**TL;DR**
I advocate to implement standards 1:1 because that's the only way to ensure compatibility across multiple implementations.
